README.md
# react-native-worry-free-style-sheet

一个省心的React-Native StyleSheet工具方法

#### 使用

`yarn add react-native-worry-free-style-sheet`

```js
import StyleSheet, { IgnoreValue } from 'react-native-worry-free-style-sheet';
const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  box1: {
    marginTop: 50,
    width: 375,
    height: new IgnoreValue(30),
    backgroundColor: 'red',
    borderWidth: 0.6,
    borderColor: 'red',
    borderRadius: 20,
    marginLeft: 10,
  },
  text: {
    fontSize: 15,
  },
});
```

#### 注意
- 如果需要忽略转换 使用 IgnoreValue 类
- 如果需要设置屏幕宽度使用你设置的基准值 默认375
- 转换结果小于一个像素的时候会返回一个像素（比如iPhone 11 width: 0.1）返回onePixel一个像素0.3333

#### 自动转换设计稿尺寸 默认 375

- 修改基准值

```js
import StyleSheet from 'react-native-worry-free-style-sheet';
StyleSheet.setConfig({
    baseWidth: 750,
});
```

#### 方法
包含默认StyleSheet的方法


| 方法/常量                    | 参数类型                               | 描述                                 |
|--------------------------|------------------------------------|------------------------------------|
| scaleSize                | number                             | 屏幕适配, 缩放size , 默认根据宽度适配，纵向也可以使用此方法 |
| setConfig                | { baseWidth }                      | 修改配置                               |
| IgnoreValue              | new IgnoreValue(100)               | 忽略转换                               |
| isIphoneX                |                                    | 判断是否为iphoneX                       |
| ifIphoneX                | iphoneXStyle,iosStyle,androidStyle | 根据是否是iPhoneX返回不同的样式                |
| statusBarHeight          |                                    | 状态栏高度                              |
| navigationBarHeight      |                                    | 常规导航栏高度（包含状态栏）                     |
| navBarHeight             |                                    | 状态栏加导航栏的高度                         |
| navBarUnlessStatusHeight |                                    | 导航栏高度(不包含状态栏)                      |
| getBottomSpace           |                                    | 获取底部边距                             |
| onePixel                 | 常量                                 | 1个像素                               |
| screenW                  | 常量                                 | 屏幕宽度                               |
| screenH                  | 常量                                 | 屏幕高度                               |
| isIOS                    | 常量                                 | 是否 iOS                             |
| isAndroid                | 常量                                 | 是否 Android                         |




#### 支持转换的样式表

```js
const transformKeys = {
    fontSize: true,
    width: true,
    height: true,
    start: true,
    end: true,
    top: true,
    left: true,
    right: true,
    bottom: true,
    minWidth: true,
    maxWidth: true,
    minHeight: true,
    maxHeight: true,
    margin: true,
    marginVertical: true,
    marginHorizontal: true,
    marginTop: true,
    marginBottom: true,
    marginLeft: true,
    marginRight: true,
    marginStart: true,
    marginEnd: true,
    padding: true,
    paddingVertical: true,
    paddingHorizontal: true,
    paddingTop: true,
    paddingBottom: true,
    paddingLeft: true,
    paddingRight: true,
    paddingStart: true,
    paddingEnd: true,
    lineHeight: true,
    borderRadius: true,
    borderWidth: true,
    borderTopWidth: true,
    borderStartWidth: true,
    borderEndWidth: true,
    borderRightWidth: true,
    borderBottomWidth: true,
    borderLeftWidth: true,
    borderBottomEndRadius: true,
    borderBottomLeftRadius: true,
    borderBottomRightRadius: true,
    borderBottomStartRadius: true,
    borderTopEndRadius: true,
    borderTopLeftRadius: true,
    borderTopRightRadius: true,
    borderTopStartRadius: true,
};
```

#### 待办
* [ ] 新增Colors常量
* [ ] 新增公共样式配置文件
* [ ] 新增中间件 方便自定义处理scaleSize转换方法
* [ ] 优化文档



CHANGELOG.md
# [1.0.0-rc.0](https://github.com/syanbo/react-native-worry-free-style-sheet/compare/v1.0.0-bata.3...v1.0.0-rc.0) (2021-04-02)

# 1.0.0-bata.3 (2021-04-02)


### Features

* add setConfig ([7d3f57a](https://github.com/syanbo/react-native-worry-free-style-sheet/commit/7d3f57ad35c32405d8fb8589de8480040dcbf72e))
* 一个省心的StyleSheet工具方法 ([7f37d72](https://github.com/syanbo/react-native-worry-free-style-sheet/commit/7f37d72aa1a759a62633f45aaf6d96642f1aabc0))
* 加入husky ([56f6295](https://github.com/syanbo/react-native-worry-free-style-sheet/commit/56f6295458931284f429074d12a1e7a2d56ae682))



src/util.ts
// @ts-ignore
import { PixelRatio, Dimensions, Platform, StatusBar } from 'react-native';

const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const STATUSBAR_HEIGHT = Platform.OS === 'ios' ? 20 : 0;
const HEADER_HEIGHT = APPBAR_HEIGHT + STATUSBAR_HEIGHT;

const getResolvedDimensions = () => {
  const { width, height } = Dimensions.get('window');
  if (width === 0 && height === 0) return Dimensions.get('screen');
  return { width, height };
};

const { width, height } = getResolvedDimensions();

export const onePixel = 1 / PixelRatio.get();
export const screenW = Math.min(width, height);
export const screenH = Math.max(width, height);

// 以iphone6为基准,如果以其他尺寸为基准的话,请修改下面的defaultWidth和defaultHeight为对应尺寸即可
const defaultWidth = 375;

// 缩放比例
let _scaleWidth = screenW / defaultWidth;

/**
 * 修改配置
 * @param baseWidth
 */
export function setConfig({ baseWidth }) {
  _scaleWidth = screenW / baseWidth;
}

/**
 * 判断是否为iphoneX
 * @returns {boolean}
 */
export const isIphoneX = () => {
  const dimen = Dimensions.get('window');
  return (
    Platform.OS === 'ios'
    && !Platform.isPad
    && !Platform.isTVOS
    && (dimen.height === 780
      || dimen.width === 780
      || dimen.height === 812 || dimen.width === 812
      || dimen.height === 844 || dimen.width === 844
      || dimen.height === 896 || dimen.width === 896
      || dimen.height === 926 || dimen.width === 926)
  );
};

/**
 * 屏幕适配, 缩放size , 默认根据宽度适配，纵向也可以使用此方法
 * @param size 设计图的尺寸
 * @returns {number}
 */
export function scaleSize(size: number) {
  // 如果是 "1%" 样式 直接返回不进行换算
  if (isNaN(size)) {
    return size;
  }
  // 如果size不为0 并且 小于一个像素 那么就返回一个像素 处理负数 - 0.1  -一个像素
  if (size && Math.abs(size) < onePixel) {
    return size < 0 ? -onePixel : onePixel;
  }
  // ok  进行换算 基准宽度 375
  const pixel = PixelRatio.getPixelSizeForLayoutSize(size) * _scaleWidth;
  return Math.floor(pixel) * onePixel;
}

/**
 * 根据是否是iPhoneX返回不同的样式
 * @param iphoneXStyle
 * @param iosStyle
 * @param androidStyle
 * @returns {*}
 */
export function ifIphoneX(iphoneXStyle, iosStyle = {}, androidStyle) {
  if (isIphoneX()) {
    return iphoneXStyle;
  } else if (Platform.OS === 'ios') {
    return iosStyle;
  } else {
    if (androidStyle) return androidStyle;
    return iosStyle;
  }
}

/**
 * 状态栏高度
 */
export function statusBarHeight() {
  if (Platform.OS === 'ios') {
    if (isIphoneX()) {
      return 44;
    }
    return 20;
  } else {
    // 自动获取导航栏高度，官方参考 720 * 1280 尺寸为 48
    return StatusBar.currentHeight;
  }
}

/**
 * 常规导航栏高度（包含状态栏）
 */
export function navigationBarHeight() {
  if (Platform.OS === 'ios') {
    if (isIphoneX()) {
      return 88;
    }
    return 64;
  } else {
    return HEADER_HEIGHT;
  }
}

/**
 * 状态栏加导航栏的高度
 * @returns {*}
 */
export function navBarHeight() {
  if (Platform.OS === 'ios') {
    if (isIphoneX()) {
      return 44 + 44;
    }
    return 20 + 44;
  } else {
    // 自动获取导航栏高度，官方参考 720 * 1280 尺寸为 48
    return StatusBar.currentHeight + 54;
  }
}

/**
 * 导航栏高度(不包含状态栏)
 * @returns {*}
 */
export function navBarUnlessStatusHeight() {
  return APPBAR_HEIGHT;
}

/**
 * 获取底部边距
 */
export function getBottomSpace() {
  return isIphoneX() ? 34 : 0;
}


src/index.ts
/**
 * Author   : Syan
 * Date     : 2019-08-27
 * Project [ godfrey ] Coded on WebStorm.
 */
// @ts-ignore
import { Platform, StyleSheet } from 'react-native';
import * as screenUtil from './util';

const { hairlineWidth, absoluteFill, flatten, absoluteFillObject, compose } = StyleSheet;
const { scaleSize } = screenUtil;

const transformKeys = {
    fontSize: true,
    width: true,
    height: true,
    start: true,
    end: true,
    top: true,
    left: true,
    right: true,
    bottom: true,
    minWidth: true,
    maxWidth: true,
    minHeight: true,
    maxHeight: true,
    margin: true,
    marginVertical: true,
    marginHorizontal: true,
    marginTop: true,
    marginBottom: true,
    marginLeft: true,
    marginRight: true,
    marginStart: true,
    marginEnd: true,
    padding: true,
    paddingVertical: true,
    paddingHorizontal: true,
    paddingTop: true,
    paddingBottom: true,
    paddingLeft: true,
    paddingRight: true,
    paddingStart: true,
    paddingEnd: true,
    lineHeight: true,
    borderRadius: true,
    borderWidth: true,
    borderTopWidth: true,
    borderStartWidth: true,
    borderEndWidth: true,
    borderRightWidth: true,
    borderBottomWidth: true,
    borderLeftWidth: true,
    borderBottomEndRadius: true,
    borderBottomLeftRadius: true,
    borderBottomRightRadius: true,
    borderBottomStartRadius: true,
    borderTopEndRadius: true,
    borderTopLeftRadius: true,
    borderTopRightRadius: true,
    borderTopStartRadius: true,
};

function transformStyle(styleObj, prop) {
    const value = styleObj[prop];
    if (!value || !transformKeys[prop]) return;
    if (value.hasOwnProperty('value')) {
        styleObj[prop] = value.value;
        return;
    }
    if (value.hasOwnProperty('ceilValue')) {
        styleObj[prop] = value.ceilValue;
        return;
    }
    if (isNaN(value)) return;
    styleObj[prop] = scaleSize(value);
}

export default {
    hairlineWidth,
    hairlineUnit: hairlineWidth,
    absoluteFill,
    absoluteFillObject,
    compose,
    flatten,
    create: StyleSheet.create((obj) => {
        for (const key in obj) {
            const styleObj = obj[key];
            if (styleObj) {
                for (const prop in styleObj) {
                    transformStyle(styleObj, prop);
                }
            }
        }
        return obj;
    }),
    ...screenUtil,
    isIOS: Platform.OS === 'ios',
    isAndroid: Platform.OS === 'android',
};

export class IgnoreValue {
    private value: number;
    constructor(value) {
        this.value = value;
    }
}

export class CeilValue {
    private ceilValue: number;
    constructor(value) {
        this.ceilValue = Math.ceil(scaleSize(value));
    }
}


src/screenUtil.ts
// @ts-ignore
import { PixelRatio, Dimensions, Platform, StatusBar } from 'react-native';

const APPBAR_HEIGHT = Platform.OS === 'ios' ? 44 : 56;
const STATUSBAR_HEIGHT = Platform.OS === 'ios' ? 20 : 0;
const HEADER_HEIGHT = APPBAR_HEIGHT + STATUSBAR_HEIGHT;

const getResolvedDimensions = () => {
  const { width, height } = Dimensions.get('window');
  if (width === 0 && height === 0) return Dimensions.get('screen');
  return { width, height };
};

const { width, height } = getResolvedDimensions();

export const onePixel = 1 / PixelRatio.get();
export const screenW = Math.min(width, height);
export const screenH = Math.max(width, height);

// 以iphone6为基准,如果以其他尺寸为基准的话,请修改下面的defaultWidth和defaultHeight为对应尺寸即可. 以下为1倍图时
const defaultWidth = 375;
const defaultHeight = 667;

// 缩放比例
const _scaleWidth = screenW / defaultWidth;
const _scaleHeight = screenH / defaultHeight;

/**
 * 判断是否为iphoneX
 * @returns {boolean}
 */
export const isIphoneX = () => {
  const dimen = Dimensions.get('window');
  return (
    Platform.OS === 'ios'
    && !Platform.isPad
    && !Platform.isTVOS
    && (dimen.height === 780
      || dimen.width === 780
      || dimen.height === 812 || dimen.width === 812
      || dimen.height === 844 || dimen.width === 844
      || dimen.height === 896 || dimen.width === 896
      || dimen.height === 926 || dimen.width === 926)
  );
};

/**
 * 屏幕适配,缩放size , 默认根据宽度适配，纵向也可以使用此方法
 * @param size 设计图的尺寸
 * @returns {number}
 */
export function scaleSize(size: number) {
  // 如果是 "1%" 样式 直接返回不进行换算
  // eslint-disable-next-line no-restricted-globals
  if (isNaN(size)) {
    return size;
  }
  // 如果size不为0 并且 小于一个像素 那么就返回一个像素 处理负数 - 0.1  -一个像素
  if (size && Math.abs(size) < onePixel) {
    return size < 0 ? -onePixel : onePixel;
  }
  // ok  进行换算 基准宽度 375
  const pixel = PixelRatio.getPixelSizeForLayoutSize(size) * _scaleWidth;
  return Math.floor(pixel) * onePixel;
}

/**
 * 屏幕适配,缩放size , 默认根据宽度适配，纵向也可以使用此方法
 * 横向的尺寸直接使用此方法
 * 如：width ,paddingHorizontal ,paddingLeft ,paddingRight ,marginHorizontal ,marginLeft ,marginRight
 * @param size 设计图的尺寸
 * @returns {number}
 */
export function scaleWidth(size: number) {
  return size * _scaleWidth;
}

/**
 * 屏幕适配 , 纵向的尺寸使用此方法应该会更趋近于设计稿
 * 如：height ,paddingVertical ,paddingTop ,paddingBottom ,marginVertical ,marginTop ,marginBottom
 * @param size 设计图的尺寸
 * @returns {number}
 */
export function scaleHeight(size: number) {
  return size * _scaleHeight;
}

/**
 * 设置字体的size（单位px）
 * @param size 传入设计稿上的px
 * @returns {Number} 返回实际sp ,会随系统缩放比例改变，如不需要请去掉 * fontScale
 */
export function setSpText(size: number) {
  const scale = Math.min(_scaleWidth, _scaleHeight);
  return size * scale;
}

/**
 * 根据是否是iPhoneX返回不同的样式
 * @param iphoneXStyle
 * @param iosStyle
 * @param androidStyle
 * @returns {*}
 */
export function ifIphoneX(iphoneXStyle, iosStyle = {}, androidStyle) {
  if (isIphoneX()) {
    return iphoneXStyle;
  } else if (Platform.OS === 'ios') {
    return iosStyle;
  } else {
    if (androidStyle) return androidStyle;
    return iosStyle;
  }
}

/**
 * 状态栏高度
 */
export function statusBarHeight() {
  if (Platform.OS === 'ios') {
    if (isIphoneX()) {
      return 44;
    }
    return 20;
  } else {
    // 自动获取导航栏高度，官方参考 720 * 1280 尺寸为 48
    return StatusBar.currentHeight;
  }
}

/**
 * 常规导航栏高度（包含状态栏）
 */
export function navigationBarHeight() {
  if (Platform.OS === 'ios') {
    if (isIphoneX()) {
      return 88;
    }
    return 64;
  } else {
    return HEADER_HEIGHT;
  }
}

/**
 * 状态栏加导航栏的高度
 * @returns {*}
 */
export function navBarHeight() {
  if (Platform.OS === 'ios') {
    if (isIphoneX()) {
      return 44 + 44;
    }
    return 20 + 44;
  } else {
    // 自动获取导航栏高度，官方参考 720 * 1280 尺寸为 48
    return StatusBar.currentHeight + 54;
  }
}

/**
 * 导航栏高度(不包含状态栏)
 * @returns {*}
 */
export function navBarUnlessStatusHeight() {
  return APPBAR_HEIGHT;
}

export function tabBarHeight() {
  if (Platform.OS === 'ios') {
    if (isIphoneX()) {
      return 88;
    }
    return 64;
  } else {
    return 55;
  }
}

export function getBottomSpace() {
  return isIphoneX() ? 34 : 0;
}
